### ISSUE TITLE
brms' GAMs smooth term not supported by emmeans

### ISSUE THREAD

**Issue by DominiqueMakowski** [paul-buerkner/brms#907]
Hi Paul,

Opening this issue following Russell's suggestion (https://github.com/rvlenth/emmeans/issues/198#issuecomment-632844645), connected to https://github.com/paul-buerkner/brms/issues/418

Essentially, the smooth term from brms models is not "extracted" (?) correctly, as it does for instance with mgcv: 

``` r
model <- mgcv::gam(Sepal.Length ~ s(Petal.Length), data=iris)
emmeans::ref_grid(model, at = list("Petal.Length"=c(2, 3, 4)))
#> 'emmGrid' object with variables:
#>     Petal.Length = 2, 3, 4

model <- brms::brm(Sepal.Length ~ s(Petal.Length), data=iris, chains=2, iter=200, refresh=0)
emmeans::ref_grid(model, at = list("Petal.Length"=c(2, 3, 4)))
#> 'emmGrid' object with variables:
#>     1 = 1
```

Russell mentioned a potential resolution: 

> if it's like the gam method, it just needs an appropriate call to a predict method (see emmeans:::emm_basis.gam)

Thanks!

**Comment by paul-buerkner** [comment]
Thanks! I take a look at `emmeans:::emm_basis.gam` and see if the same thing can be done for brms.

**Comment by paul-buerkner** [comment]
To clarify, smooths are not extracted at all, which is expected behavior. Right now, intentionally, only "fixed effects" are extracted so we have in a way minimal support for emmeans as of now. This may be extended in the future in various directions.

**Comment by DominiqueMakowski** [comment]
I'm not sure I understand, do you mean that the 'issue' is located in emmeans in `emmeans:::emm_basis.gam`? Or that allowing the extraction and subsequent taking into account of the smooth term has to be performed in brms? In any case, thanks for your help!

**Comment by paul-buerkner** [comment]
I just mean I need to see if I can copy over the behavior of `emmeans:::emm_basis.gam` to the corresponding `brmsfit` method.

**Comment by DominiqueMakowski** [comment]
@paul-buerkner really sorry for the bump but would you have any rough idea when you'd have the time to look into it?  As I'd need this feature for some analysis, it would help plan out a timeline ☺️ and, if you don't no worries, maybe there's something I can help out with?

**Comment by paul-buerkner** [comment]
I am sorry, I don't. If you want to help you could of course check how the
.gam method does it and provide ideas how to get it working in brms.

Dominique Makowski <notifications@github.com> schrieb am Sa., 4. Juli 2020,
12:43:

**Comment by DominiqueMakowski** [comment]
No worries,

Here's the code for [`emm_basis.gam`](https://github.com/rvlenth/emmeans/blob/master/R/gam-support.R#L76):

```r
emm_basis.gam = function(object, trms, xlev, grid, 
                         freq = FALSE, unconditional = FALSE, ...) {
    # coef() works right for lm but coef.aov tosses out NAs
    bhat = object$coefficients
#    m = suppressWarnings(model.frame(trms, grid, na.action = na.pass, xlev = xlev))
#    X = model.matrix(trms, m, contrasts.arg = object$contrasts)
    X = predict(object, newdata = grid, type = "lpmatrix", 
                newdata.guaranteed = TRUE)
    bhat = as.numeric(bhat) 
    # stretches it out if multivariate - see mlm method
    V = .my.vcov(object, freq = freq, unconditional = unconditional, ...)
    
    # if (sum(is.na(bhat)) > 0)
    #     nbasis = estimability::nonest.basis(object$qr)
    # else
        nbasis = estimability::all.estble
        misc = .std.link.labels(object$family, list())
        # dffun = function(k, dfargs) Inf
        # dfargs = list()
#    else {
        dfargs = list(df = object$df.residual)
        dffun = function(k, dfargs) dfargs$df
#    }
    list(X=X, bhat=bhat, nbasis=nbasis, V=V, dffun=dffun, dfargs=dfargs, misc=misc)
}
```

However, I'm not sure where to look for adding it in brms, I've searched the [`brmsfit-methods`](https://github.com/paul-buerkner/brms/blob/master/R/brmsfit-methods.R) but didn't manage where the GAM part is. Could you guide me a bit so that I can try adding it?

**Comment by mattansb** [comment]
Just to note that this issue isn't just localized to `emm_basis`, but also to `recover_data`, where the `terms` object generated here:

https://github.com/paul-buerkner/brms/blob/master/R/emmeans.R#L33

Does not include the smoothed predictor's term(s), only the "fixed effects" per se, as Paul has already mentioned (:

**Comment by DominiqueMakowski** [comment]
Trying to clarify the issue

Looking at `recover_data`:

https://github.com/paul-buerkner/brms/blob/203cc81775b327b9de6e6d93d1aa720e149889c7/R/emmeans.R#L38-L45

and [emm_basis.brmsfit](https://github.com/paul-buerkner/brms/blob/203cc81775b327b9de6e6d93d1aa720e149889c7/R/emmeans.R#L51), it seems like there is a common occurrence of ` .extract_par_terms(object, resp, dpar, nlpar)`:

https://github.com/paul-buerkner/brms/blob/203cc81775b327b9de6e6d93d1aa720e149889c7/R/emmeans.R#L60

Would returning the smooth term in here fix this issue? Or is it the wrong place to look for the solution

(@mattansb sorry 🙏  to ask your help again but you're the best brms&emmeans expert I know ^^)

**Comment by mattansb** [comment]
Seem like in `recover_data()` the issue can be resolved by using `bterms$allvars`:

```R
recover_data.brmsfit <- function (object, data, resp = NULL, dpar = NULL, nlpar = NULL, ...) {
  bterms <- brms:::.extract_par_terms(object, resp, dpar, nlpar)
  trms <- attr(model.frame(bterms$allvars, data = object$data), "terms")
  emmeans::recover_data(call("brms"), trms, "na.omit", object$data, ...)
}
```

In `emm_basis` be have the issue that smooth terms' parameters don't start with `"b_"`, and there are more than 1...

I do wonder if all of that code can be replaced with `posterior_linpred()`?

@paul-buerkner What do you think of this?

```R
emm_basis.brmsfit <- function (object, trms, xlev, grid, vcov., resp = NULL, dpar = NULL, nlpar = NULL, ...) {
  if (brms:::is_equal(dpar, "mean")) {
    post.beta <- rstantools::posterior_epred(object, newdata = grid, 
                                             re_formula = NA, ...)
    misc <- list()
  }
  else {
    bterms <- brms:::.extract_par_terms(object, resp, dpar, nlpar)
    post.beta <- rstantools::posterior_linpred(object, newdata = grid, 
                                               resp = resp,
                                               dpar = dpar,
                                               nlpar = nlpar,
                                               re_formula = NA, ...)
    misc <- emmeans::.std.link.labels(bterms$family, list())
  }
  bhat <- apply(post.beta, 2, mean)
  V <- cov(post.beta)
  X <- diag(length(bhat))
  attr(post.beta, "n.chains") <- object$fit@sim$chains
  nbasis <- matrix(NA)
  dfargs <- list()
  dffun <- function(k, dfargs) Inf
  environment(dffun) <- baseenv()
  brms:::nlist(X, bhat, nbasis, V, dffun, dfargs, misc, post.beta)
}


model <- brms::brm(Sepal.Length ~ s(Petal.Length), data=iris, chains=2, iter=200, refresh=0)

emmeans::ref_grid(model, at = list("Petal.Length"=c(2, 3, 4)))
#> 'emmGrid' object with variables:
#>     Petal.Length = 2, 3, 4
```

Just to compare the actual emmeans:

```R
emmeans::emmeans(model, ~Petal.Length, at = list("Petal.Length"=c(2, 3, 4)))
#>  Petal.Length emmean lower.HPD upper.HPD
#>             2   5.17      4.96      5.38
#>             3   5.20      4.86      5.45
#>             4   5.74      5.63      5.88
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95

model2 <- mgcv::gam(Sepal.Length ~ s(Petal.Length), data=iris)

emmeans::emmeans(model2, ~Petal.Length, at = list("Petal.Length"=c(2, 3, 4)))
#>  Petal.Length emmean     SE  df lower.CL upper.CL
#>             2   5.13 0.1276 143     4.88     5.38
#>             3   5.12 0.1675 143     4.79     5.45
#>             4   5.70 0.0738 143     5.56     5.85
#> 
#> Confidence level used: 0.95 
```

**Comment by paul-buerkner** [comment]
If this would work, that would be absolutely perfect! I still don't fully understand all the requirements of emmeans. @rvlenth would you mind taking a quick look if this approach is something we can do in general?

**Comment by mattansb** [comment]
Also seems to work when margnializing over other predictors:

```R
model <- brms::brm(Sepal.Length ~ s(Petal.Length) + Species + Sepal.Width, data=iris, chains=2, iter=200, refresh=0)
(em <- emmeans::emmeans(model, ~Petal.Length, at = list("Petal.Length"=c(2, 3, 4))))
#>  Petal.Length emmean lower.HPD upper.HPD
#>             2   4.62      4.27      5.11
#>             3   5.24      4.94      5.44
#>             4   5.93      5.55      6.11
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95

model2 <- mgcv::gam(Sepal.Length ~ s(Petal.Length) + Species + Sepal.Width, data=iris)
(em2 <- emmeans::emmeans(model2, ~Petal.Length, at = list("Petal.Length"=c(2, 3, 4))))
#>  Petal.Length emmean     SE  df lower.CL upper.CL
#>             2   4.65 0.1570 144     4.33     4.96
#>             3   5.25 0.0819 144     5.08     5.41
#>             4   5.90 0.0968 144     5.71     6.09
#> 
#> Results are averaged over the levels of: Species 
#> Confidence level used: 0.95 
```

But for some reason we're not getting the `Results are averaged over the levels of: Species` footnote for `brms`...

The only issue I can see with this method is that the `@linfct` slot does not match the model's parameters, but instead it matches the produced predictors:

```R
em@linfct
#>           [,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]      [,8]      [,9]
#> [1,] 0.3333333 0.3333333 0.3333333 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000
#> [2,] 0.0000000 0.0000000 0.0000000 0.3333333 0.3333333 0.3333333 0.0000000 0.0000000 0.0000000
#> [3,] 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.3333333 0.3333333 0.3333333

em2@linfct
#>      (Intercept) Speciesversicolor Speciesvirginica Sepal.Width s(Petal.Length).1 s(Petal.Length).2 s(Petal.Length).3 s(Petal.Length).4 s(Petal.Length).5
#> [1,]           1         0.3333333        0.3333333    3.057333         1.1240806         0.3641188        -0.8392458        0.43020446        0.85007794
#> [2,]           1         0.3333333        0.3333333    3.057333         0.8699502         0.0488611        -0.1232841        0.82295454        0.01223351
#> [3,]           1         0.3333333        0.3333333    3.057333         0.1210959        -0.3857765         0.1851327        0.03331455        0.19587948
#>      s(Petal.Length).6 s(Petal.Length).7 s(Petal.Length).8 s(Petal.Length).9
#> [1,]         0.6245913        -0.3896401        -0.1534668        -0.9992020
#> [2,]         0.6586687        -0.1876355         0.4444958        -0.4308277
#> [3,]         0.2666129         0.2009198         0.5153090         0.1375466
```

**Comment by DominiqueMakowski** [comment]
I have added support for that in `estimate_slopes()`, which is based on `emtrends` so it should be pretty general:

``` r
model <- mgcv::gam(Petal.Length ~ s(Sepal.Width), data = iris)
x <- modelbased::estimate_slopes(model, modulate = "Sepal.Width", length = 30)
plot(modelbased::visualisation_recipe(x))
```

![](https://i.imgur.com/x98y2Is.png)

``` r
model <- lm(Petal.Length ~ poly(Sepal.Width, 4), data = iris)
x <- modelbased::estimate_slopes(model, modulate = "Sepal.Width", length = 30)
plot(modelbased::visualisation_recipe(x))
```

![](https://i.imgur.com/DeeXlPC.png)

<sup>Created on 2021-06-08 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>



Regarding brms support, I'll wait a bit to see if Mat manages to be awesome again :p

**Comment by DominiqueMakowski** [comment]
(oops I commented and closed the wrong issue)

**Comment by rvlenth** [comment]
Well, failing to correctly match the model parameters is a huge reason to think this isn't right. The `linfct` slot for `em` above does not include an intercept and seems to just equally average together some indicators and not account for the smooth. It looks like apples and oranges to me. 

What we need to get for `emmeans` support is not terribly mysterious:

  1. In `recover_data`, we need to recover all the variables in the data used for fitting the model that pertain to predicting the fixed-effect  parameters for the specified part of the model. (in the example above, that include Petal.Length, Species, and Sepal.Width.) If there is smoothing, there are usually "specials" in the terms() part of the model that need to be preserved in the `trms` argument.
  2. In `emm_basis` we need to construct an appropriate model matrix given a specified grid of predictor settings (goes into `X`), along with the corresponding regression coefficients (in this case the posterior sample thereof (goes into `post.beta`, from which you can calculate `bhat` and `V` for use in frequentist summaries, if requested).)

In `gam` models, I can satisfy (2) by taking advantage of its "lpmatrix" option in `predict.gam`. It sets up the spline basis needed to make the predictions I want.

Other than saying that, I don't have any specific advice because I don't know much about what **brms** does internally.

**Comment by mattansb** [comment]
> The linfct slot for em above does not include an intercept and seems to just equally average together some indicators and not account for the smooth.

@rvlenth The `grid` passed to `posterior_linpred` has input values for all the terms in the model (including the smooth). It seems (@paul-buerkner can you confirm?) that `posterior_linpred` internally builds the model matrix (accounting for the smooth, intercept, any other predictors, etc...) and then produces posteriors of the linear combination.

So having an `emm_basis` function that is based on this method is like returning an `emmGrid` that has been `regrid`ed. See example:

```R
model <- brms::brm(Sepal.Length ~ s(Petal.Length), data=iris, chains=2, iter=200, refresh=0)
em <- emmeans::emmeans(model, ~Petal.Length, at = list("Petal.Length"=c(2, 4, 8)))

model2 <- mgcv::gam(Sepal.Length ~ s(Petal.Length), data=iris)
em2 <- emmeans::emmeans(model2, ~Petal.Length, at = list("Petal.Length"=c(2, 4, 8)))

emmeans::emmip(model, ~Petal.Length, cov.red = unique, CIs = TRUE)
```
![image](https://user-images.githubusercontent.com/35330040/121243549-8071cc80-c8a6-11eb-876e-2faab113f632.png)

```R
emmeans::emmip(model2, ~Petal.Length, cov.red = unique, CIs = TRUE)
```
![image](https://user-images.githubusercontent.com/35330040/121243587-8a93cb00-c8a6-11eb-825e-429b062abeed.png)


```R
emmeans::regrid(em)@V
#>               [,1]          [,2]         [,3]
#> [1,]  0.0142476450 -0.0002303544 -0.011109973
#> [2,] -0.0002303544  0.0052217513 -0.002022907
#> [3,] -0.0111099727 -0.0020229065  0.354893825
emmeans::regrid(em2)@V
#>               [,1]          [,2]         [,3]
#> [1,]  0.0162893028 -0.0002116121 -0.005525305
#> [2,] -0.0002116121  0.0054533443 -0.001478784
#> [3,] -0.0055253049 -0.0014787835  0.504088993

emmeans::regrid(em)@bhat
#> [1] 5.11526 5.70927 8.85507
emmeans::regrid(em2)@bhat
#> [1] 5.131039 5.703430 8.713981
```

Can you elaborate what is missing for this solution to be an acceptable method to add emmeans support to brms?

**Comment by rvlenth** [comment]
I don't have more to say other than that the `linfct`s looked very different and have different numbers of columns. I can only comment on what I am shown. And I am not sure why the graphs I received by email differ from those I see on this page.

Top graph from my email:
![image](https://user-images.githubusercontent.com/26777519/121248382-feaf7e00-c868-11eb-966d-53505d98b320.png)

Top graph from above:
![image](https://user-images.githubusercontent.com/26777519/121248447-1424a800-c869-11eb-9205-9107b2e51d4a.png)

These differ substantially!

**Comment by mattansb** [comment]
Sorry about that, I went back and edited my response to use a more complex example, and updated the graphs etc.

As I noted, the number of columns in linfct comply with the number of predictions generated by `posterior_linpred` and not the number of coefficients - this is simply a different parameterization of the emmeans. I think I've demonstrated that this method produces equivalent results using a different parameterization.

@rvlenth @paul-buerkner let me know if I can be of any more help (:

**Comment by paul-buerkner** [comment]
Thank you @mattansb and @rvlenth for both of your insights and explanation! I am now convinced that this new approach is doing the right thing and in a very elegant and general manner. `posterior_epred` and `posterior_linpred` are very closely related in that they return predictions per observation and posterior draw, only that the former always gives us the mean of the posterior predictive distribution while the latter gives us the latent linear (additive) predictor of a particular predicted parameter in the model. In simple cases, `posterior_epred = inv_link(posterior_linpred)`. In other words, if the `dpar = "mean"` approach worked before (and I have no doubt it does), so will the new `posterior_linpred` approach. I will make a PR in which I implement this change.

**Comment by mattansb** [comment]
Fantastic!
I would add a note to the docs explaining the odd linfact slot, perhaps referencing back to this issue for more details.

Thanks!

**Comment by paul-buerkner** [comment]
Implemented via #1177
